src/radical/pilot/pmgr/launching/base.py

```diff
@@ -25,34 +123,1121 @@ def __init__(self, cfg, session):
+
+        # register input queue
+        self.register_input(rps.PMGR_LAUNCHING_PENDING,
+                            rpc.PMGR_LAUNCHING_QUEUE, self.work)
+
+        # we don't really have an output queue, as we pass control over the
+        # pilot jobs to the resource management system (ResourceManager).
+
+        self._stager_queue = self.get_output_ep(rpc.STAGER_REQUEST_QUEUE)
+
+        # we listen for pilot cancel commands
+        self.register_subscriber(rpc.CONTROL_PUBSUB, self._pmgr_control_cb)
+
+        # also listen for completed staging directives
+        self.register_subscriber(rpc.STAGER_RESPONSE_PUBSUB, self._staging_ack_cb)
+        self._active_sds = dict()
+        self._sds_lock   = mt.Lock()
+
+        self._log.info(ru.get_version([self._mod_dir, self._root_dir]))
+        self._rp_version, _, _, _, self._rp_sdist_name, self._rp_sdist_path = \
+                ru.get_version([self._mod_dir, self._root_dir])
+
+
+        # load all launcher implementations
+        self._launchers = list()
+
+        from .saga  import PilotLauncherSAGA
+        from .psi_j import PilotLauncherPSIJ
+
+        impl = {
+            RP_UL_NAME_SAGA : PilotLauncherSAGA,
+            RP_UL_NAME_PSI_J: PilotLauncherPSIJ
+        }
+
+        for name in [RP_UL_NAME_PSI_J, RP_UL_NAME_SAGA]:
+            try:
+                self._launchers.append(impl[name](name, self._log, self._prof,
+                    self._state_cb))
+            except:
+                self._log.exception('skip launcher %s' % name)
+
 
     # --------------------------------------------------------------------------
     #
-    # This class-method creates the appropriate sub-class for the Launcher.
-    #
     @classmethod
     def create(cls, cfg, session):
 
-        name = cfg.get('pmgr_launching_component', RP_UL_NAME_DEFAULT)
+        return cls(cfg, session)
 
-        # Make sure that we are the base-class!
-        if cls != PMGRLaunchingComponent:
-            raise TypeError("Factory only available to base class!")
 
-        from .default import Default
+    # --------------------------------------------------------------------------
+    #
+    def _state_cb(self, pilot, rp_state):
 
-        impl = {
-            RP_UL_NAME_DEFAULT: Default
-        }
+        self._log.info('=== pilot state update: %s: %s', pilot['uid'], rp_state)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def finalize(self):
+
+        # FIXME: is this called?
+
+        try:
+            self.unregister_input(rps.PMGR_LAUNCHING_PENDING,
+                                  rpc.PMGR_LAUNCHING_QUEUE, self.work)
+
+            # FIXME: always kill all pilot jobs for non-final pilots at
+            #        termination, and set the pilot states to CANCELED.
+            with self._pilots_lock:
+                pids = list(self._pilots.keys())
+
+            self._cancel_pilots(pids)
+            self._kill_pilots(pids)
```

Comment:
`_cancel_pilots` is called within `_kill_pilots`, should it be removed either from here or from `_kill_pilots`?

Reply:
ack, done.